src/shared/backend/graph/firmware/cloudbuild.ts

```diff
@@ -0,0 +1,199 @@
+import config from "shared/backend/config";
+import { createBuilder } from "shared/backend/utils/builder";
+
+const builder = createBuilder();
+
+// Targets data
+
+const Release = builder.simpleObject("Release", {
+  fields: (t) => ({
+    id: t.string(),
+    name: t.string(),
+    description: t.string({ nullable: true }),
+    timestamp: t.string(),
+    excludeTargets: t.stringList(),
+    sha: t.string(),
+    isPrerelease: t.boolean(),
+  }),
+});
+
+const Flag = builder.simpleObject("Flag", {
+  fields: (t) => ({
+    id: t.string(),
+    values: t.stringList(),
+  }),
+});
+
+const Target = builder.simpleObject("Target", {
+  fields: (t) => ({
+    id: t.string(),
+    name: t.string(),
+    tags: t.stringList(),
+  }),
+});
+
+const TagFlag = builder.simpleObject("TagFlag", {
+  fields: (t) => ({
+    id: t.string(),
+    values: t.stringList(),
+  }),
+});
+
+const Tag = builder.simpleObject("Tag", {
+  fields: (t) => ({
+    id: t.string(),
+    tagFlags: t.field({ type: [TagFlag] }),
+  }),
+});
+
+const CloudTargets = builder.simpleObject("CloudTargets", {
+  fields: (t) => ({
+    releases: t.field({ type: [Release] }),
+    targets: t.field({ type: [Target] }),
+    flags: t.field({ type: [Flag] }),
+    tags: t.field({ type: [Tag] }),
+  }),
+});
+
+// Job status
+
+const CloudFirmware = builder.simpleObject("CloudFirmware", {
+  fields: (t) => ({
+    base64Data: t.string(),
+  }),
+});
+
+const CloudFirmwareStatus = builder.simpleObject("CloudFirmwareStatus", {
+  fields: (t) => ({
+    status: t.string(),
+    downloadUrl: t.string({ nullable: true }),
+  }),
+});
+
+const CloudFirmwareParams = builder.inputType("CloudFirmwareParams", {
+  fields: (t) => ({
+    release: t.string({ required: true }),
+    target: t.string({ required: true }),
+    flags: t.field({
+      type: [
+        builder.inputType("SelectedFlag", {
+          fields: (t__) => ({
+            name: t__.string({ required: true }),
+            value: t__.string({ required: true }),
+          }),
+        }),
+      ],
+      required: true,
+    }),
+  }),
+});
+
+builder.queryType({
+  fields: (t) => ({
+    cloudTargets: t.field({
+      type: CloudTargets,
+      resolve: async (_, __, { cloudbuild, github }) => {
+        const cloudTargets = await cloudbuild.fetchTargets();
+        const firmwaresReleases = new Set(Object.keys(cloudTargets.releases));
+        const githubReleases = await github(
+          "GET /repos/{owner}/{repo}/releases",
+          {
+            owner: config.github.organization,
+            repo: config.github.repos.firmware,
+          }
+        );
+
```

Comment:
Minor point, but you should consider making the `github` and `cloudbuild` calls in a `Promise.all` to increase query speed. They don't depend on each other

Reply:
Valid point, thanks for the info.